### ISSUE TITLE
extended-metadata not added as JSON

### ISSUE THREAD

**Issue by nshttpd** [terraform-providers/terraform-provider-oci#817]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version and Provider Version

```
Terraform v0.11.14
+ provider.oci v3.32.0
```

### Affected Resource(s)

`oci_core_instance`

### Terraform Configuration Files

```hcl
resource "oci_core_instance" "test-instance" {
  availability_domain = "${var.ad_list[1]}"
  compartment_id      = "${var.compartment_id}"
  shape               = "VM.Standard1.2"

  agent_config {
    is_monitoring_disabled = "false"
  }

  create_vnic_details {
    subnet_id        = "${var.node_subnets[1]}"
    assign_public_ip = "false"
  }

  display_name = "test-instance"

  extended_metadata = {
    extra-logs = "[\"/var/log/commerce/pci.log\", \"/var/log/commerce/trasaction.log\"]"
  }

  metadata = {
    ssh_authorized_keys = "${var.ssh_public_key}"
    app-name            = "extra-logs-test"
  }

  source_details {
    source_id   = "ocid1.image.oc1.phx.aaaaaaaa5b5tbfa4qkmu5fkm2m4aaluaqu73f32peylcjhs3vaglu6e223yq"
    source_type = "image"
  }

  preserve_boot_volume = false
}
```

### Expected Behavior

The extended-metadata value for the `extra-logs` key should be a JSON Array stored in the metadata. When using the OCI CLI to update the instance in such a way as :

`oci compute instance update --instance-id ocid1.instance.oc1.phx.OCID --extended-metadata file://./extended-md.json`

where the extended-md.json file contains :

```
{
  "extra-logs": ["/var/log/commerce/pci.log", "/var/log/commerce/trasaction.log"]
}
```
the value for `extra-logs` in the metadata is the JSON array as expected.

```
{
  "availabilityDomain": "PHX-AD-2",
  "faultDomain": "FAULT-DOMAIN-2",
  "compartmentId": "ocid1.compartment.oc1..COMPARTMENT",
  "displayName": "test-instance",
  "id": "ocid1.instance.oc1.phx.OCID",
  "image": "ocid1.image.oc1.phx.aaaaaaaa5b5tbfa4qkmu5fkm2m4aaluaqu73f32peylcjhs3vaglu6e223yq",
  "metadata": {
    "extra-logs": [
      "/var/log/commerce/pci.log",
      "/var/log/commerce/trasaction.log"
    ],
    "app-name": "extra-logs-test",
    "ssh_authorized_keys": "ssh-rsa MAHBUCKET"
  },
  "region": "phx",
  "canonicalRegionName": "us-phoenix-1",
  "shape": "VM.Standard1.2",
  "state": "Running",
  "timeCreated": 1562359164500,
  "agentConfig": {
    "monitoringDisabled": false
  }
}
```

### Actual Behavior

This is what the metadata looks like on the instance with the extended-metadata being applied to the instance via the Terraform Provider.

```
{
  "availabilityDomain": "PHX-AD-2",
  "faultDomain": "FAULT-DOMAIN-2",
  "compartmentId": "ocid1.compartment.oc1..COMPARTMENT",
  "displayName": "test-instance",
  "id": "ocid1.instance.oc1.phx.OCID",
  "image": "ocid1.image.oc1.phx.aaaaaaaa5b5tbfa4qkmu5fkm2m4aaluaqu73f32peylcjhs3vaglu6e223yq",
  "metadata": {
    "extra-logs": "[\"/var/log/commerce/pci.log\", \"/var/log/commerce/trasaction.log\"]",
    "app-name": "extra-logs-test",
    "ssh_authorized_keys": "ssh-rsa MAHBUCKET"
  },
  "region": "phx",
  "canonicalRegionName": "us-phoenix-1",
  "shape": "VM.Standard1.2",
  "state": "Running",
  "timeCreated": 1562359164500,
  "agentConfig": {
    "monitoringDisabled": false
  }
}
```

### Steps to Reproduce

Provision an instance using the OCI Provider and follow the example instructions in the documentation of how to supply a JSON blob as a value for extended-metadata value.

1. `terraform apply`

### References

A quick look through the Provider code seems to possibly have some weird logic in `core_instance_resource.go` on L1154 in `mapToExtendedMetadata`. Trying to Unmarshal the string value from the TF file into `map[string]interface{}` will fail for this use case and just store out the original string. Just a guess, I've been staring at a lot of code this week and it's all starting to blur together.

**Comment by alexng-canuck** [comment]
This is a limitation of the Terraform provider's `mapToExtendedMetadata` function.

The `json.Unmarshal` call is unmarshaling the JSON array string into a map type; which results in an error. As a result, the provider falls back to using the literal JSON string when passing the `extended_metadata` to service.

As a workaround, the array can be wrapped inside a JSON map to avoid this.

We should however investigate a fix for this so that other JSON types can be unmarshalled. Perhaps coercing the unmarshal result into an `interface{}` type instead of a `map[string]interface{}` type.

**Comment by nshttpd** [comment]
Yes, after more experimentation, instead of making the Value for the extended-metadata a JSON array changing it to pushing all the Configuration into a single Key with a JSON Map as the value containing all the application configuration information I wanted to have worked. It is an inconsistency with how the OCI CLI handles extended-metadata, but knowing the limitation is there allows an acceptable workaround.

**Comment by nshttpd** [comment]
@alexng-canuck not sure if you want to leave this open for future investigation at some point or close it since it's not really a bug, but missing functionality compared to the CLI.

**Comment by parrneet** [comment]
Fixed in Release 3.34.0
